datafusion/functions/src/datetime/date_bin.rs

```diff
@@ -481,17 +656,82 @@ fn date_bin_impl(
                         origin, stride, stride_fn, array, tz_opt,
                     )?
                 }
+                Time32(Millisecond) => {
```

Comment:
The new match arms' bodies are almost identical. A helper function/macro could be extracted and reused.